lib/color_spinor_field.cpp

```diff
@@ -876,19 +882,697 @@ namespace quda {
+    }
+
+    if (ghost_field_reset) destroyIPCComms();
+    createIPCComms();
+  }
+
+  // pack the ghost zone into a contiguous buffer for communications
+  void ColorSpinorField::packGhost(const int nFace, const QudaParity parity, const int dagger,
+                                   const qudaStream_t &stream, MemoryLocation location[2 * QUDA_MAX_DIM],
+                                   MemoryLocation location_label, bool spin_project, double a, double b, double c,
+                                   int shmem)
+  {
+    if (Location() == QUDA_CPU_FIELD_LOCATION) errorQuda("Host field not supported");
+    void *packBuffer[4 * QUDA_MAX_DIM] = {};
+
+    for (int dim=0; dim<4; dim++) {
+      for (int dir=0; dir<2; dir++) {
+        switch (location[2 * dim + dir]) {
+
+        case Device: // pack to local device buffer
+          packBuffer[2 * dim + dir] = my_face_dim_dir_d[bufferIndex][dim][dir];
+          packBuffer[2 * QUDA_MAX_DIM + 2 * dim + dir] = nullptr;
+          break;
+        case Shmem:
+          // this is the remote buffer when using shmem ...
+          // if the ghost_remote_send_buffer_d exists we can directly use it
+          // - else we need pack locally and send data to the recv buffer
+          packBuffer[2 * dim + dir] = ghost_remote_send_buffer_d[bufferIndex][dim][dir] != nullptr ?
+            static_cast<char *>(ghost_remote_send_buffer_d[bufferIndex][dim][dir]) + ghost_offset[dim][1 - dir] :
+            my_face_dim_dir_d[bufferIndex][dim][dir];
+          packBuffer[2 * QUDA_MAX_DIM + 2 * dim + dir] = ghost_remote_send_buffer_d[bufferIndex][dim][dir] != nullptr ?
+            nullptr :
+            static_cast<char *>(ghost_recv_buffer_d[bufferIndex]) + ghost_offset[dim][1 - dir];
+          break;
+	case Host:   // pack to zero-copy memory
+	  packBuffer[2*dim+dir] = my_face_dim_dir_hd[bufferIndex][dim][dir];
+          break;
+        case Remote: // pack to remote peer memory
+          packBuffer[2 * dim + dir]
+            = static_cast<char *>(ghost_remote_send_buffer_d[bufferIndex][dim][dir]) + ghost_offset[dim][1 - dir];
+          break;
+	default: errorQuda("Undefined location %d", location[2*dim+dir]);
+        }
+      }
     }
-    return fine;
+
+    PackGhost(packBuffer, *this, location_label, nFace, dagger, parity, spin_project, a, b, c, shmem, stream);
+  }
+
+  void ColorSpinorField::pack(int nFace, int parity, int dagger, const qudaStream_t &stream,
+                                  MemoryLocation location[2 * QUDA_MAX_DIM], MemoryLocation location_label,
+                                  bool spin_project, double a, double b, double c, int shmem)
+  {
+    if (Location() == QUDA_CPU_FIELD_LOCATION) errorQuda("Host field not supported");
+    createComms(nFace, spin_project); // must call this first
+
+    packGhost(nFace, (QudaParity)parity, dagger, stream, location, location_label, spin_project, a, b, c, shmem);
+  }
+
+  // FIXME reconcile with above
```

Comment:
for this PR?

Reply:
I consider the packing aspects as part of the comms cleanup, which as we've discussed, is deferred for a follow up PR